### PR TITLE
Updated 3rd class from RO 16th Anniversary Patch

### PR DESCRIPTION
**Pull request by aleos89**
* **Addressed Issue(s)**: #3602, #3653

* **Server Mode**: Renewal

* **Description of Pull Request**: 
  * Updated the 3rd class skills according to the RO 16th Anniversary Patch.
  * More info here: http://ro.gnjoy.com/news/devnote/View.asp?category=1&seq=3890090&curpage=1

There are a couple of TODO to confirm some remaining things. They are commented where info is needed.

### CODE REVIEW

**Comment by attackjom** [comment]
When weapon blocking has activated the "counter state". I can't walk or do something for a while around 10 secs.

**Comment by aleos89** [comment]
Please read the link I posted with all of the skill changes.

**Comment by Badarosk0** [comment]
Ai﻿med Bolt
(Number of attack won't depen﻿d on the immobilize status of target.)
Don't work.

**Comment by Badarosk0** [comment]
* Spore Explosion - Wrong - In the images the skill generates a total of 3 hits. You are currently only generating 1 hit.

* Counter Slash - Wrong - Can be used without any prerequisites. It is to be able to spam only in 10 seconds of Counter State when active when blocking a skill.

* Fatal Manace - Wrong - When used in conjunction with a Dagger, it is for 2 hits. Generated only 1 hit with dagger.

**Comment by zackdreaver** [comment]
[kRO 20180827 SC.zip](https://github.com/rathena/rathena/files/2568082/kRO.20180827.SC.zip)
Fatal Manace in action

**Comment by zackdreaver** [comment]
There is new EFST_COUNTER_ON in stateiconinfo
적에게 공격 명중 시확률에 따라 카운터차기 준비 자세

and according to skill desc:
While Weapon Blocking is active and player successfuly blocked a melee physical attack,
player will enter **counter state** that enabling the use of **Counter Slash** for 10 seconds,

My guess, without actual ingame proof:
During 10 seconds of **counter state**:
- player can do anything (like moving, attacking, using skill/item)
- player can use **Counter Slash** only **ONCE**, after **counter state** is cancelled

**Comment by OptimusM** [comment]
i was testing  Tetra Vortex skill, and i was looking alot about dmg formula, in the formula says it will increase the dmg by 200MATK every lvl over 5, but actually i was testing and the dmg increases from 70k on lvl 5 to 140k at lvl 10.

i tried to gather more info about it but according to this gif, the splash at lvl 10 do less dmg than tetra at lvl 5

http://prntscr.com/lkzwj8

**Comment by OptimusM** [comment]
@aleos89 i tried the fix, but still works like i said, thx for take a look of that

**Comment by aleos89** [comment]
Should be good now. I realized the mistake I made.

**Comment by OptimusM** [comment]
Now it´s working Perfectly, Thank you @aleos89

File: src/map/battle.cpp [hunk 1]
```diff
@@ -5630,8 +5621,6 @@ static struct Damage battle_calc_weapon_attack(struct block_list *src, struct bl
 		case CR_ACIDDEMONSTRATION:
 		case GN_FIRE_EXPANSION_ACID:
 #endif
-		case SO_VARETYR_SPEAR:
-			return wd; //These skills will do a GVG fix later
```
Comment (Victor214) [inline src/map/battle.cpp]: Is this `return wd;` line supposed to be removed aswell?
Wouldn't it affect the other skills above it?

**Comment by OptimusM** [comment]
idk if this just happen to me but in some cases Tetra Vortex skill bring´s back the black square on hit, i mean like this
https://github.com/rathena/rathena/issues/1353

Weird thing, just happen when you use "Release"
or when you cast Stone or Wind element (this does not happen when u cast normaly using water and fire)

Anyone can relate this?

**Comment by Badarosk0** [comment]
While Weapon Blocking is active and player successfuly blocked a melee physical attack,
player will enter counter state that enabling the use of Counter Slash for 10 seconds,

https://youtu.be/UugsqMvAiXM?t=46

Follow the video showing the behavior. @aleos89

**Comment by Badarosk0** [comment]
Skill: GN_SPORE_EXPLOSION

Currently in the rA, if the player uses a card that increases the long damage, it will be increased.

I researched this skill and everything says it is melee and not long atk.

http://kafra.kr/#!/en/KRO/skillinfo/2481
```
스포어 익스플로젼
MAX Lv : 10
습득조건 : 스페셜 파머시 4 
계열 :   특수 / 데미지 
내용 :  대상 1 체에게 폭발성 버섯 포자를 붙인다. 2초 후  버섯 포자가 폭발하면서, 대상 및 주변에 물리 데미지를 준다.
시전자의 BaseLv 및 INT에 따라 데미지가 증가되며, 스킬 사용시 [폭발 버섯 포자] 를 하나 소모한다. 
[Level 1] :  공격력 350% / 효과 범위 3 x 3 셀
[Level 2] :  공격력 500% / 효과 범위 3 x 3 셀
[Level 3] :  공격력 650% / 효과 범위 5 x 5 셀
[Level 4] :  공격력 800% / 효과 범위 5 x 5 셀
[Level 5] :  공격력 950% / 효과 범위 7 x 7 셀
[Level 6] :  공격력 1100% / 효과 범위 7 x 7 셀
[Level 7] :  공격력 1250% / 효과 범위 9 x 9 셀
[Level 8] :  공격력 1400% / 효과 범위 9 x 9 셀
[Level 9] :  공격력 1550% / 효과 범위 11 x 11 셀
[Level10] :  공격력 1700% / 효과 범위 11 x 11 셀
```
```
Spore Explosion
MAX Lv: 10
Prerequisite: Special Farmer 4
Category: Special / Damage
Description: Attach explosive mushroom spores to a target. After 2 seconds, the mushroom spores explode, giving physical damage to the target and surrounding area.
Increases the damage done by the caster's BaseLv and INT, and consumes one [Explosive Mushroom Spore] when using Skill.
[Level 1]: Attack power 350% / Effect range 3 x 3 cells
[Level 2]: Damage 500% / Effect range 3 x 3 cells
[Level 3]: Attack power 650% / Effect range 5 x 5 cells
[Level 4]: Damage 800% / Effect range 5 x 5 cells
[Level 5]: Attack power 950% / Effect range 7 x 7 cells
[Level 6]: Damage 1100% / Effect range 7 x 7 cells
[Level 7]: Attack power 1250% / Effect range 9 x 9 cells
[Level 8]: Attack power 1400% / Effective range 9 x 9 cells
[Level 9]: Attack power 1550% / Effect range 11 x 11 cells
[Level 10]: Attack power 1700% / Effect range 11 x 11 cells
```
@aleos89

**Comment by aleos89** [comment]
Would need to confirm this on kRO, but the reason is here:
```cpp
static int battle_range_type(struct block_list *src, struct block_list *target, uint16 skill_id, uint16 skill_lv)
...
	//based on used skill's range
	if (skill_get_range2(src, skill_id, skill_lv, true) < 4)
		return BF_SHORT;
	return BF_LONG;
```

**Comment by UniverseCodes** [comment]
Question, is this episode 17.1 or no

**Comment by cydh** [comment]
that GN_SPORE_EXPLOSION maybe because 11 is the "throw" range, and the damage itself must be melee (imagine about sticky bomb), and also the splash area as melee.
but while the default skill_attack which trigger battle calc still use that throw range which is not correct. need some workaround

**Comment by cahya1992** [comment]
But i asking question about that skill to sigma and he said its that skill is long ranged
https://www.divine-pride.net/forum/index.php?/topic/3280-16th-anniversary-update-note-part-1/&do=findComment&comment=6096

and this is the new formula for GN_SPORE_EXPLOSION
https://www.divine-pride.net/forum/index.php?/topic/3280-16th-anniversary-update-note-part-1/&do=findComment&comment=6100

**Comment by ecdarreola** [comment]
Once, sc_weaponblock_on is triggered, character is not allowed to move?

**Comment by attackjom** [comment]
When use Fire Tetravortex via the "Release" skill  it still have black image.

**Comment by Lokylro** [comment]
Good job ! The skills work perfectly. There is just AB judex skill that is incorrect at the distance of use ... is it an official change?

**Comment by laziem** [comment]
```Program received signal SIGSEGV, Segmentation fault.
0x081d7bb0 in skill_check_condition_castbegin (sd=<optimized out>, skill_id=<optimized out>,
    skill_lv=<optimized out>) at skill.cpp:16108
16108                           if (!sc->data[SC_WEAPONBLOCK_ON]) {
(gdb) bt full
#0  0x081d7bb0 in skill_check_condition_castbegin (sd=<optimized out>,
    skill_id=<optimized out>, skill_lv=<optimized out>) at skill.cpp:16108
        status = 0xf35ba74
        sc = <optimized out>
        require = {hp = 0, mhp = 0, sp = 27, hp_rate = 0, sp_rate = 0, zeny = 0, weapon = 0,
          ammo = 0, ammo_qty = 0 '\000', state = 0 '\000', spiritball = 0 '\000', itemid = {0,
            0, 0, 0, 0, 0, 0, 0, 0, 0}, amount = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0}, eqItem = 0x0,
          status = 0xac5225ac, status_count = 1 '\001', eqItem_count = 0 '\000'}
        i = <optimized out>
        inf2 = <optimized out>
        inf3 = <optimized out>
#1  0x08245908 in unit_skilluse_id2 (src=0xf35b728, target_id=2014131, skill_id=2029,
    skill_lv=10, casttime=0, castcancel=0) at unit.cpp:1709
        ud = 0xf35b73c
        tstatus = 0xf35ba74
        sc = <optimized out>
        sd = 0xf35b728
        target = 0xf35b728
        tick = <optimized out>
        combo = 0
        range = <optimized out>
        inf = 40 '('
        inf2 = 255179380
#2  0x0824652a in unit_skilluse_id (src=0xf35b728, target_id=2014131, skill_id=2029,
    skill_lv=10) at unit.cpp:1334
No locals.
#3  0x080c64a6 in clif_parse_skill_toid (sd=0xf35b728, skill_id=2029, skill_lv=<optimized out>,
    target_id=2014131) at clif.cpp:12691
        tick = <optimized out>
        inf = <optimized out>
#4  0x080c686e in clif_parse_UseSkillToId (fd=23, sd=0xf35b728) at clif.cpp:12701
        info = <optimized out>
#5  0x080b2497 in clif_parse (fd=23) at clif.cpp:21504
        cmd = 1080
        packet_len = 10
        sd = 0xf35b728
        pnum = 0
#6  0x08051a76 in main ()
No symbol table info available.
```

Crash from WEAPONBLOCK_ON

**Comment by aleos89** [comment]
`SC_WEAPONBLOCK_ON` doesn't exist in `skill_check_condition_castbegin`. It's in `skill_check_condition_sc_required`, so maybe you merged incorrectly?

**Comment by laziem** [comment]
Ahhh.. My bad, yes, turn out I merge incorrectly...

**Comment by nehpetskie** [comment]
> `SC_WEAPONBLOCK_ON` doesn't exist in `skill_check_condition_castbegin`. It's in `skill_check_condition_sc_required`, so maybe you merged incorrectly?

i want to ask about magnus = magnus hit only UNDEAD AND DEMON monster in the patch its declare all monster and 30% more dmg to undead and demon

**Comment by nehpetskie** [comment]
ALSO RUNE UPDATE ?? in kro no changes? in EFFECT and skill DB sir aleos89 c

**Comment by aleos89** [comment]
Magnus Exorcismus is not changed in this PR, this is for 3rd class changes only.

I'm not sure what you're implying for the RK changes.

**Comment by nehpetskie** [comment]
> Magnus Exorcismus is not changed in this PR, this is for 3rd class changes only.
> 
> I'm not sure what you're implying for the RK changes.

Master aleos HERE
http://ro.gnjoy.com/News/devnote/view.asp?seq=3978525


Im talking about the PR_MAGNUS about that pull you made here at 
https://github.com/rathena/rathena/pull/4072/commits/8accb49669ad3024aa23ed88deaf1bc5cf644563?file-filters%5B%5D=.hpp

thank you for replying ^_^

**Comment by aleos89** [comment]
These Rune Knight changes are not part of this update. The link you provided was something done a year after these changes. I'll keep it separate from this pull request.

**Comment by nehpetskie** [comment]
> These Rune Knight changes are not part of this update. The link you provided was something done a year after these changes. I'll keep it separate from this pull request.

ok copy kindly reply on my? pm on discord. im skie21

**Comment by AceWindstorm** [comment]
Hi, I came across this and wanted to help out. I began researching and writing up information, but before I do it for all skills that need verification I would just like to post what I have so far to make sure it is sufficient. Once complete I can simply upload a text file with all the information, rather than copypasting a large amount like I am about to for Genesis Ray below.




LG_RAYOFGENESIS :

Video - https://www.youtube.com/watch?v=jauuoDfUBu0

Video date - 9/29/18

Update date - 8/27/18

Gear in Video :

Amistr Beret
Magical Booster
Toy Syringe
Robust Magical Armor
Old Parasol
Bloody Knight Shield/Purified Knight Shield (cant tell really)
Fallen Angel Wings
Temporal Int Boots
Illusion Battle Chips L and R

Illusion Battle Chips cannot be enchanted to reduce aftercast or skill cooldown as far as I can tell. Source : https://www.divine-pride.net/forum/index.php?/topic/3324-kro-episode-171-enchant-info-illusion-equipment-and-os-weapons/

The Knight Shields require +12 to get aftercast delay reduction, his is +9 so none there.

Robust Magical Armor cant be enchanted at all let alone with aftercast as far as I can tell from my research.

Amistr Beret affects fixed casting time by .1 sec per refine above 10. His is +11, so .1 sec.

Magical Booster gives 1% aftercast delay reduction, effectively nothing and will be unnoticable.

All other gear for sure gives no bonus that would interfere, the above items were ones whose effects I did not know and therefore listed as proof and to better give context.


From what I can tell using a stopwatch, the cast time is approximately .2 seconds, meaning that it is a .3 fixed cast time normally. Genesis Ray also appears to have a 2 second cooldown, with 1 second of aftercast delay.





If this is insufficient let me know or if you can point out a better/more accurate method then please do so!

@aleos89

**Comment by laziem** [comment]
Bug report:
Spore Explosion explode immediately, should be 2 sec after cast. Changing value in skill_cast_db.txt didn't have any effect

**Comment by attackjom** [comment]
So confuse about damage type of spore explosion

skilldescription is mention this skill inflicts long range physical damage.

**Comment by laziem** [comment]
Update: Spore Explosion still immediately explode

**Comment by Badarosk0** [comment]
> So confuse about damage type of spore explosion
> 
> skilldescription is mention this skill inflicts long range physical damage.

```
[SKID.GN_SPORE_EXPLOSION] = {
		"스포어 익스플로젼",
		"MAX Lv : 10",
		"^777777습득조건 : 스페셜 파머시 4 ^000000",
		"계열 : ^777777  액티브 / 데미지  ^000000",
		"내용 : ^777777 대상 1 체에게 폭발성 버섯 포자를 붙여 1차적으로 대상에게 원거리 물리 데미지를 준 뒤, 폭발시켜 2차적으로 대상 및 주변에 원거리 물리 데미지를 준다.",
		"1차 데미지는 시전자의 BaseLv 및 INT에 따라 데미지가 증가되며, 2차 데미지는 시전자의 BaseLv에 따라 데미지가 증가된다.",
		"스킬 사용시 [폭발 버섯 포자] 를 하나 소모한다. ^000000",
		"[Level 1] : ^777777 공격력 380%/180%/범위: 3x3셀^000000",
		"[Level 2] : ^777777 공격력 560%/360%/범위: 3x3셀^000000",
		"[Level 3] : ^777777 공격력 740%/540%/범위: 5x5셀^000000",
		"[Level 4] : ^777777 공격력 920%/720%/범위: 5x5셀^000000",
		"[Level 5] : ^777777 공격력1100%/900%/범위: 7x7셀^000000",
		"[Level 6] : ^777777 공격력1280%/1080%/범위: 7x7셀^000000",
		"[Level 7] : ^777777 공격력1460%/1260%/범위: 9x9셀^000000",
		"[Level 8] : ^777777 공격력1640%/1440%/범위: 9x9셀^000000",
		"[Level 9] : ^777777 공격력1820%/1620%/범위:11x11셀^000000",
		"[Level10] : ^777777 공격력2000%/1800%/범위:11x11셀^000000"
	},
```
"google translator" :
```
Spore Explosion,
"MAX Lv: 10",
"^ 777777 Requirement: Special Farmer's 4 ^ 000000",
"Series: ^ 777777 active / damage ^ 000000",
"Description: ^ 777777 Attach 1 explosive mushroom spore to a target, dealing long ranged physical damage to the target first, then exploding and dealing long ranged physical damage to the target and surroundings.",
"Damaged damage increases with the caster's BaseLv and INT, and secondary damage increases with the caster's BaseLv.",
"Skill consumes 1 [Explosive Mushroom Spore]. ^ 000000",
"[Level 1]: ^ 777777 Damage 380% / 180% / Range: 3x3 cells ^ 000000",
"[Level 2]: ^ 777777 Damage 560% / 360% / Range: 3x3 cells ^ 000000",
"[Level 3]: ^ 777777 Damage 740% / 540% / Range: 5x5 cells ^ 000000",
"[Level 4]: ^ 777777 Damage 920% / 720% / Range: 5x5 cells ^ 000000",
"[Level 5]: ^ 777777 Damage1100% / 900% / Range: 7x7 cells ^ 000000",
"[Level 6]: ^ 777777 Damage1280% / 1080% / Range: 7x7 cells ^ 000000",
"[Level 7]: ^ 777777 Damage 1460% / 1260% / Range: 9x9 cells ^ 000000",
"[Level 8]: ^ 777777 Damage 1640% / 1440% / Range: 9x9 cells ^ 000000",
"[Level 9]: ^ 777777 Damage1820% / 1620% / Range: 11x11 cells ^ 000000",
"[Level10]: ^ 777777 Damage 2000% / 1800% / Range: 11x11 cells ^ 000000"
```

**Comment by Badarosk0** [comment]
There is a problem with the Varetyr Spear skill being cast next to the target and not considering the range.

**Comment by whatsofunny** [comment]
Reviewed, Spore Explosion still immediately explode even there's a countdown.

**Comment by aleos89** [comment]
I’m going to ignore the Spore Explosion delay because the next batch of updates for Genetics removes the delay.

**Comment by teededung** [comment]
Spore Explosion still immediately explode
Solve by adding `NoDamage: true` in `DamageFlags:`
@aleos89 @laziem @whatsofunny

**Pull request merged**